libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts

```diff
@@ -414,6 +422,54 @@ export class WorkspaceInstallManager extends BaseInstallManager {
 
     // TODO: Remove when "rush link" and "rush unlink" are deprecated
     LastLinkFlagFactory.getCommonTempFlag(this.rushConfiguration).create();
+
+    // Stage 2 rebuild pending
+    if (this.twoStageInstall && !this.options.ignoreScripts) {
+      const rebuildArgs: string[] = ['rebuild', '--pending', '--recursive'];
+
+      if (
+        this.rushConfiguration.pnpmOptions.pnpmStore === 'local' ||
+        EnvironmentConfiguration.pnpmStorePathOverride
+      ) {
+        // --store is not supported to pnpm rebuild... So, config.store is used here
+        rebuildArgs.push(`--config.storeDir=${this.rushConfiguration.pnpmOptions.pnpmStorePath}`);
+      }
+
+      if (this.options.collectLogFile) {
+        rebuildArgs.push('--reporter', 'ndjson');
+      }
+
+      for (const arg of this.options.pnpmFilterArguments) {
+        rebuildArgs.push(arg);
+      }
+
+      // If any diagnostic options were specified, then show the full command-line
+      if (this.options.debug || this.options.collectLogFile || this.options.networkConcurrency) {
+        console.log(
+          os.EOL +
+            colors.green('Invoking package manager: ') +
+            FileSystem.getRealPath(this._packageManagerFilename) +
+            ' ' +
+            rebuildArgs.join(' ') +
+            os.EOL
+        );
+      }
+
+      try {
+        Utilities.executeCommand({
+          command: this._packageManagerFilename,
+          args: rebuildArgs,
+          workingDirectory: this.rushConfiguration.commonTempFolder,
+          environment: this._packageManagerEnv,
+          suppressOutput: false
+        });
+      } catch (err) {
+        throw new Error(`Encounter an error when running install lifecycle scripts. error: ${err.message}`);
+      } finally {
+        // Always save after pnpm rebuild to update timestamp of last install flag file.
+        this.commonTempInstallFlag.create();
```

Comment:
Previously, this flag would only exist if there was a successful install. Use should be updated here to keep in line with what I describe above (update the flag for `ignoreScripts` in the case that the post-install scripts run + succeed, or leave as-is and use it to skip the install but use to execute post-install scripts.

Reply:
The reason why invoking creating last install flag here: `canSkipInstall` compares a timestamp(mtime of this flag file) with mtime of node_modules folder. And no matter `pnpm rebuild --pending` runs successfully or not, `node_modules` folder is touched, i.e. mtime updated. So we need to update the mtime of this flag file to make the `canSkipInstall` works again.

Reply:
Right, but skipping an install in the event of an error doesn't make sense. It should be updated only after the command is successfully executed. If the command fails, then the state of the repo is unknown and the last-install flag should probably be deleted.

Reply:
It's kindof fight against the main point of the proposal, faster retries on postinstall scripts.

Imagine

```
pnpm install --ignore-scripts --filter <a...>
```

works

```
pnpm rebuild --pending --filter <a...>
```

If `pnpm rebuild` fails, Rush.js doesn't update the last-install.flag. The "canSkipInstall" will never return true in this case. which triggers a "pnpm install --ignore-scripts --filter <a...>" again...

but, the main point is to make "canSkipInstall" returns true in this case, and only the stage 2 command("pnpm rebuild" ) runs, which is the "faster retires" part...

Reply:
Then in that case it should write the flag file twice, once after performing the install (with `ignoreScripts: true`), then once after running the post-install scripts (with `ignoreScripts: false`). We shouldn't be saying "yes, everything is installed correctly" regardless of success or failure, that defeats the entire purpose of the flag file.

Reply:
> in that case it should write the flag file twice, once after performing the install

+1  Rush install is meant to be transactionally safe.  It is a common optimization for CI agents to reuse the same `git clone` working folder from a previous job. And it is very common for the previous `rush install` to have been terminated because someone pushed an update to their PR branch, which causes the previous CI job to get immediately cancelled and restarted.